### PR TITLE
Fix for S3 plugin to specify correct region parameter for us-east-1. …

### PR DESCRIPTION
**Pull request by debarshiraha**
Fix for #16978

### CODE REVIEW

**Comment by karmi** [comment]
Hi @debarshiraha, we have found your signature in our records, but it seems like you have signed with a different e-mail than the one used in yout Git [commit](https://github.com/elastic/elasticsearch/pull/20689.patch). Can you please add both of these e-mails into your Github profile (they can be hidden), so we can match your e-mails to your Github profile?

**Comment by debarshiraha** [comment]
Hi @karmi, I've added another e-mail, but still the same issue. I see that I've signed using the same e-mail that I've with github. Do we know which email needs to be added? Or should I send you my github e-mail?

**Comment by debarshiraha** [comment]
Looks like CLA check has passed now!

**Comment by dadoonet** [comment]
Thanks for this fix.
TBH I wonder if we should also open an issue in AWS SDK? I mean that their client should support that. May be they fixed that in a recent update?

To me your fix is a temporary fix. We could also fix it within the documentation instead.

**Comment by debarshiraha** [comment]
I understand, but this is a known issue, and the logic is as per the API. The latest SDK (1.11.39/40) doesn't incorporate this at the SDK level as well. I also found this https://github.com/boto/boto3/issues/125 and as per the SDK developer (for boto, but applies the same for Java I guess): "We actually dynamically generate methods at runtime based on a shared JSON Description. So this essentially means that we surface the exact REST API in boto3 (and all the semantics of the underlying REST API)".

So this looks the best fix for me as per the API, unless we want to customize this fix in the SDK itself somehow.

I  did not get the documentation fix proposal.

**Comment by dadoonet** [comment]
Hi @debarshiraha 

> I did not get the documentation fix proposal.

I meant that we can modify simply modify https://github.com/elastic/elasticsearch/blob/master/docs/plugins/repository-s3.asciidoc and add that for `us-east-1` region, `region` must not be set. And then we don't need to fix the code.

BTW I looked at AWS SDK and apparently they merged that in: https://github.com/aws/aws-sdk-java/pull/758/files

I can see in the code now: https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/model/Region.java#L261-L263

```java
        if (s3RegionId == null || s3RegionId.equals("US") || s3RegionId.equals("us-east-1")) {
            return Region.US_Standard;
        }
```

So basically if we instead update AWS SDK to `1.11.36` (1.11.37 has been released in the meantime), we should be ok. WDYT?

**Comment by debarshiraha** [comment]
I see. I already tested on a newer version (1.11.39/40) with this change in and this basic code still fails:
```new AmazonS3Client().createBucket(new CreateBucketRequest("bucket", "us-east-1"));```

So we can try with the newer version, but not sure if this Region enum is being used here.

For the doc change, as of now not specifying the region doesn't work as well.

**Comment by debarshiraha** [comment]
Another option is that we change the code to 

```new CreateBucketRequest(bucket, com.amazonaws.services.s3.model.Region.fromValue(region)))```

And update the SDK version. So looks like some code change is required (not sure if just doc update is enough/best experience for the user). WDYT? How do we test the SDK version update (to test for any jar/dependency issue etc)?

**Comment by dadoonet** [comment]
I like [your last proposal](https://github.com/elastic/elasticsearch/pull/20689#issuecomment-250687929). It makes sense to me.
But as you wrote, updating to the last SDK version is not that easy. 

I started that work a while ago with #19594 but was blocked at some point by https://github.com/aws/aws-sdk-java/issues/788 which will not be fixed apparently on AWS side. So it needs some code changes as well with the security manager.

**Comment by debarshiraha** [comment]
Checking back, are we going to update the sdk version soon?

**Comment by dadoonet** [comment]
@debarshiraha If it's needed by your fix, then I guess you can upgrade the AWS SDK within your PR. But you have to change some other stuff if you do that as I wrote in https://github.com/elastic/elasticsearch/pull/20689#issuecomment-250691874. I'd do that within 2 different commits:

* First one: upgrade AWS SDK
* Second one:  fix the code for your issue

WDYT?

**Comment by dadoonet** [comment]
@debarshiraha Do you think you would come with an update for your PR? Thanks!

**Comment by dadoonet** [comment]
No further news on this one. And we are probably going to remove the `region` setting to keep only `endpoint` as we discussed in #22758.

So I'm closing but thanks for the effort you made @debarshiraha.
